web/src/hashchange.js

```diff
@@ -56,6 +57,7 @@ function show_all_message_view() {
     const coming_from_inbox = maybe_hide_inbox();
     const is_actively_scrolling = message_scroll.is_actively_scrolling();
     narrow.deactivate(!(coming_from_recent_view || coming_from_inbox), is_actively_scrolling);
+    left_sidebar_navigation_area.highlight_all_messages_view();
```

Comment:
This should almost certainly be done inside `narrow.deactivate` itself, via `handle_post_view_change`.

We're trying to make `deactivate` no longer exist, instead just being a set of parameters passed to `activate`; so I think the right mechanism here is just that `left_sidebar_navigation_area.handle_narrow_activated` should do this when the filter is "all messages".

Reply:
I think this block of code is where that function attempts to this do this, improperly; so probably that `if` statement should just call `highlight_all_messages_view` instead of doing a hacky version of it by hand?

```
    ops = filter.operands("in");                                                                        
    if (ops.length >= 1) {                                                                              
        filter_name = ops[0];                                                                           
        if (filter_name === "home") {                                                                   
            $filter_li = $(".top_left_all_messages");                                                   
            $filter_li.addClass("active-filter");                                                       
        }                                                                                               
    }                                                                                                   
```

Reply:
Hmmmm. The limitation to that code extends from its containing `handle_narrow_activated()` function, which does not get called on a reload.

While the code I wrote in `hashchange.js` is adjacent a call to `narrow.deactivate`, I'm not actually doing anything with that...instead just calling `highlight_all_messages_view()`, since the containing `show_all_message_view()` function was the only thing I could easily identify as firing when the All message view is reloaded.

Reply:
Also, I just tested `handle_post_view_change()`, and that does not appear to get called when the all-messages view is reloaded.

Reply:
I would expect it to be called -- `show_all_message_view` calls `narrow.deactivate`, which calls `handle_post_view_change`. Aha, here's the bug1 `narrow.deactivate` is being passed the wrong value for its first parameter, `coming_from_all_messages`; we're coming from nowhere.

```
function show_all_message_view() {                                                                      
    const coming_from_recent_view = maybe_hide_recent_view();                                           
    const coming_from_inbox = maybe_hide_inbox();                                                       
    const is_actively_scrolling = message_scroll.is_actively_scrolling();                               
    narrow.deactivate(!(coming_from_recent_view || coming_from_inbox), is_actively_scrolling);          
    // We need to maybe scroll to the selected message                                                  
    // once we have the proper viewport set up                                                          
    setTimeout(message_viewport.maybe_scroll_to_selected, 0);                                           
}                                                                                                       
```

The right fix for this starts with #23812 (part of that refactoring will be making `message_lists.current` undefined when we haven't gone somewhere yet, but I think directly doing that change is unsafe this late in a release cycle). I think the best workaround is to just add a `narrow.has_visited_all_messages` global and making this change:

```
    // Both All messages and Recent Conversations have `undefined` filter.                              
    // Return if already in the All message narrow.                                                     
+    if (narrow_state.filter() === undefined && coming_from_all_messages && has_visited_all_messages) {                              
        return;                                                                                         
    }                                                                                                   
+  has_visited_all_messages = true;
```

Since that's localized and easy to undo and will fix the on-reload bug, which still using the intended `handle_post_view_change` code path for notifying the rest of the app about view changes and thus keeping us making progress towards #23812.

Reply:
Cool. I just pushed a reworked commit that I believe implements the fix you're describing here. Let me know if I've missed the mark, or if there are things to improve. Thanks, @timabbott!